File: test/language/offsets/java/offsets/NestedOffsetTest.java
package offsets;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import offsets.nested_offset.NestedOffset;
import offsets.nested_offset.NestedOffsetArrayStructure;
import offsets.nested_offset.NestedOffsetChoice;
import offsets.nested_offset.NestedOffsetStructure;
import offsets.nested_offset.NestedOffsetUnion;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.BitStreamWriter;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.SerializeUtil;

public class NestedOffsetTest
{
    @Test
    public void readConstructor() throws IOException, ZserioError
    {
        final boolean writeWrongOffsets = false;
        final BitBuffer bitBuffer = writeNestedOffsetToBitBuffer(writeWrongOffsets);
        final BitStreamReader reader = new ByteArrayBitStreamReader(bitBuffer);
        final NestedOffset nestedOffset = new NestedOffset(reader);
        checkNestedOffset(nestedOffset);
    }

    @Test
    public void readConstructorWrongOffsets() throws IOException, ZserioError
    {
        final boolean writeWrongOffsets = true;
        final BitBuffer bitBuffer = writeNestedOffsetToBitBuffer(writeWrongOffsets);
        final BitStreamReader reader = new ByteArrayBitStreamReader(bitBuffer);
        assertThrows(ZserioError.class, () -> new NestedOffset(reader));
    }

    @Test
    public void bitSizeOf()
    {
        final boolean createWrongOffsets = false;
        final NestedOffset nestedOffset = createNestedOffset(createWrongOffsets);
        assertEquals(NEST_OFFSET_BIT_SIZE, nestedOffset.bitSizeOf());
    }

    @Test
    public void bitSizeOfWithPosition()
    {
        final boolean createWrongOffsets = false;
        final NestedOffset nestedOffset = createNestedOffset(createWrongOffsets);
        final int bitPosition = 2;
        assertEquals(NEST_OFFSET_BIT_SIZE - bitPosition, nestedOffset.bitSizeOf(bitPosition));
    }

    @Test
    public void initializeOffsets()
    {
        final boolean createWrongOffsets = true;
        final NestedOffset nestedOffset = createNestedOffset(createWrongOffsets);
        final int bitPosition = 0;
        assertEquals(NEST_OFFSET_BIT_SIZE, nestedOffset.initializeOffsets(bitPosition));
        checkNestedOffset(nestedOffset);
    }

    @Test
    public void initializeOffsetsWithPosition()
    {
        final boolean createWrongOffsets = true;
        final NestedOffset nestedOffset = createNestedOffset(createWrongOffsets);
        final int bitPosition = 2;
        assertEquals(NEST_OFFSET_BIT_SIZE , nestedOffset.initializeOffsets(bitPosition));
        checkNestedOffset(nestedOffset);
    }

    @Test
    public void writeRead() throws IOException, ZserioError
    {
        final boolean createWrongOffsets = false;
        final NestedOffset nestedOffset = createNestedOffset(createWrongOffsets);
        final BitBuffer bitBuffer = SerializeUtil.serialize(nestedOffset);
        checkNestedOffset(nestedOffset);

        final NestedOffset readNestedOffset = SerializeUtil.deserialize(NestedOffset.class, bitBuffer);
        checkNestedOffset(readNestedOffset);
        assertTrue(nestedOffset.equals(readNestedOffset));
    }

    @Test
    public void writeReadWithPosition() throws IOException, ZserioError
    {
        final boolean createWrongOffsets = true;
        final NestedOffset nestedOffset = createNestedOffset(createWrongOffsets);
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        final int bitPosition = 2;
        writer.writeBits(0, bitPosition);
        nestedOffset.initializeOffsets(writer.getBitPosition());
        nestedOffset.write(writer);
        checkNestedOffset(nestedOffset);

        final BitStreamReader reader = new ByteArrayBitStreamReader(
                writer.toByteArray(), writer.getBitPosition());
        assertEquals(0, reader.readBits(bitPosition));
        final NestedOffset readNestedOffset = new NestedOffset(reader);
        checkNestedOffset(readNestedOffset);
        assertTrue(nestedOffset.equals(readNestedOffset));
    }

    @Test
    public void writeWrongOffsets() throws ZserioError, IOException
    {
        final boolean createWrongOffsets = true;
        final NestedOffset nestedOffset = createNestedOffset(createWrongOffsets);
        final BitStreamWriter writer = new ByteArrayBitStreamWriter();
        assertThrows(ZserioError.class, () -> nestedOffset.write(writer));
        writer.close();
    }

    private BitBuffer writeNestedOffsetToBitBuffer(boolean writeWrongOffsets) throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            writer.writeUnsignedInt((writeWrongOffsets) ? WRONG_TERMINATOR_OFFSET : TERMINATOR_OFFSET);
            writer.writeBool(BOOL_VALUE);
            writer.writeVarSize(NestedOffsetUnion.CHOICE_nestedOffsetArrayStructure); // union's choice tag
            writer.writeUnsignedByte(NUM_ELEMENTS);
            for (short i = 0; i < NUM_ELEMENTS; ++i)
            {
                writer.writeUnsignedInt((writeWrongOffsets) ? WRONG_DATA_OFFSET : FIRST_DATA_OFFSET + i * 8L);
                writer.writeBits(0, (i == 0) ? 7 : 1);
                writer.writeBits(i, 31);
            }

            writer.alignTo(8);
            writer.writeBits(TERMINATOR_VALUE, 7);

            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }

    private void checkNestedOffset(NestedOffset nestedOffset)
    {
        assertEquals(TERMINATOR_OFFSET, nestedOffset.getTerminatorOffset());
        assertEquals(BOOL_VALUE, nestedOffset.getBoolValue());

        final NestedOffsetChoice nestedOffsetChoice = nestedOffset.getNestedOffsetChoice();
        assertEquals(BOOL_VALUE, nestedOffsetChoice.getType());

        final NestedOffsetUnion nestedOffsetUnion = nestedOffsetChoice.getNestedOffsetUnion();
        assertEquals(NestedOffsetUnion.CHOICE_nestedOffsetArrayStructure, nestedOffsetUnion.choiceTag());

        final NestedOffsetArrayStructure nestedOffsetArrayStructure =
                nestedOffsetUnion.getNestedOffsetArrayStructure();
        assertEquals(NUM_ELEMENTS, nestedOffsetArrayStructure.getNumElements());

        final NestedOffsetStructure[] nestedOffsetStructureList =
                nestedOffsetArrayStructure.getNestedOffsetStructureList();
        assertEquals(NUM_ELEMENTS, nestedOffsetStructureList.length);
        for (short i = 0; i < NUM_ELEMENTS; ++i)
        {
            final NestedOffsetStructure nestedOffsetStructure = nestedOffsetStructureList[i];
            assertEquals(FIRST_DATA_OFFSET + i * 8L, nestedOffsetStructure.getDataOffset());
            assertEquals(i, nestedOffsetStructure.getData());
        }

        assertEquals(TERMINATOR_VALUE, nestedOffset.getTerminator());
    }

    private NestedOffset createNestedOffset(boolean createWrongOffsets)
    {
        final NestedOffsetStructure[] nestedOffsetStructureList = new NestedOffsetStructure[NUM_ELEMENTS];
        for (short i = 0; i < NUM_ELEMENTS; ++i)
        {
            final long dataOffset = (createWrongOffsets) ? WRONG_DATA_OFFSET : FIRST_DATA_OFFSET + i * 8L;
            nestedOffsetStructureList[i] = new NestedOffsetStructure(dataOffset, i);
        }

        final NestedOffsetArrayStructure nestedOffsetArrayStructure =
                new NestedOffsetArrayStructure(NUM_ELEMENTS, nestedOffsetStructureList);

        final NestedOffsetUnion nestedOffsetUnion = new NestedOffsetUnion();
        nestedOffsetUnion.setNestedOffsetArrayStructure(nestedOffsetArrayStructure);

        final NestedOffsetChoice nestedOffsetChoice = new NestedOffsetChoice(BOOL_VALUE);
        nestedOffsetChoice.setNestedOffsetUnion(nestedOffsetUnion);

        final long terminatorOffset = (createWrongOffsets) ? WRONG_TERMINATOR_OFFSET : TERMINATOR_OFFSET;
        final NestedOffset nestedOffset = new NestedOffset(terminatorOffset, BOOL_VALUE, nestedOffsetChoice,
                TERMINATOR_VALUE);

        return nestedOffset;
    }

    private static final boolean BOOL_VALUE = true;
    private static final short NUM_ELEMENTS = 2;

    private static final long WRONG_TERMINATOR_OFFSET = 0;
    private static final long TERMINATOR_OFFSET = 7 + NUM_ELEMENTS * 8;

    private static final long WRONG_DATA_OFFSET = 0;
    private static final long FIRST_DATA_OFFSET = 7 + 4;

    private static final byte TERMINATOR_VALUE = 0x45;

    private static final int NEST_OFFSET_BIT_SIZE = (int)(TERMINATOR_OFFSET * 8 + 7);
}

File: test/extensions/language/offsets/python/NestedOffsetTest.py
import zserio

import Offsets


class NestedOffsetTest(Offsets.TestCase):
    def testBitSizeOf(self):
        createWrongOffsets = False
        nestedOffset = self._createNestedOffset(createWrongOffsets)
        self.assertEqual(self.NEST_OFFSET_BIT_SIZE, nestedOffset.bitsizeof())

    def testBitSizeOfWithPosition(self):
        createWrongOffsets = False
        nestedOffset = self._createNestedOffset(createWrongOffsets)
        bitPosition = 2
        self.assertEqual(self.NEST_OFFSET_BIT_SIZE - bitPosition, nestedOffset.bitsizeof(bitPosition))

    def testInitializeOffsets(self):
        createWrongOffsets = True
        nestedOffset = self._createNestedOffset(createWrongOffsets)
        bitPosition = 0
        self.assertEqual(self.NEST_OFFSET_BIT_SIZE, nestedOffset.initialize_offsets(bitPosition))
        self._checkNestedOffset(nestedOffset)

    def testInitializeOffsetsWithPosition(self):
        createWrongOffsets = True
        nestedOffset = self._createNestedOffset(createWrongOffsets)
        bitPosition = 2
        self.assertEqual(self.NEST_OFFSET_BIT_SIZE, nestedOffset.initialize_offsets(bitPosition))
        self._checkNestedOffset(nestedOffset)

    def testRead(self):
        writeWrongOffsets = False
        writer = zserio.BitStreamWriter()
        self._writeNestedOffsetToStream(writer, writeWrongOffsets)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        nestedOffset = self.api.NestedOffset.from_reader(reader)
        self._checkNestedOffset(nestedOffset)

    def testReadWrongOffsets(self):
        writeWrongOffsets = True
        writer = zserio.BitStreamWriter()
        self._writeNestedOffsetToStream(writer, writeWrongOffsets)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        with self.assertRaises(zserio.PythonRuntimeException):
            self.api.NestedOffset.from_reader(reader)

    def testWrite(self):
        createWrongOffsets = True
        nestedOffset = self._createNestedOffset(createWrongOffsets)
        writer = zserio.BitStreamWriter()
        nestedOffset.initialize_offsets(writer.bitposition)
        nestedOffset.write(writer)
        self._checkNestedOffset(nestedOffset)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readNestedOffset = self.api.NestedOffset.from_reader(reader)
        self._checkNestedOffset(readNestedOffset)
        self.assertTrue(nestedOffset == readNestedOffset)

    def testWriteWithPosition(self):
        createWrongOffsets = True
        nestedOffset = self._createNestedOffset(createWrongOffsets)
        writer = zserio.BitStreamWriter()
        bitPosition = 2
        writer.write_bits(0, bitPosition)
        nestedOffset.initialize_offsets(writer.bitposition)
        nestedOffset.write(writer)
        self._checkNestedOffset(nestedOffset)

    def testWriteWrongOffsets(self):
        createWrongOffsets = True
        nestedOffset = self._createNestedOffset(createWrongOffsets)
        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            nestedOffset.write(writer)

    def _writeNestedOffsetToStream(self, writer, writeWrongOffsets):
        writer.write_bits(self.WRONG_TERMINATOR_OFFSET if writeWrongOffsets else self.TERMINATOR_OFFSET, 32)
        writer.write_bool(self.BOOL_VALUE)
        writer.write_varsize(self.api.NestedOffsetUnion.CHOICE_NESTED_OFFSET_ARRAY_STRUCTURE)

        writer.write_bits(self.NUM_ELEMENTS, 8)
        for i in range(self.NUM_ELEMENTS):
            writer.write_bits(
                self.WRONG_DATA_OFFSET if writeWrongOffsets else self.FIRST_DATA_OFFSET + i * 8, 32
            )
            writer.write_bits(0, 7 if (i == 0) else 1)
            writer.write_bits(i, 31)

        writer.alignto(8)
        writer.write_bits(self.TERMINATOR_VALUE, 7)

    def _checkNestedOffset(self, nestedOffset):
        self.assertEqual(self.TERMINATOR_OFFSET, nestedOffset.terminator_offset)
        self.assertEqual(self.BOOL_VALUE, nestedOffset.bool_value)

        nestedOffsetChoice = nestedOffset.nested_offset_choice
        self.assertEqual(self.BOOL_VALUE, nestedOffsetChoice.type)

        nestedOffsetUnion = nestedOffsetChoice.nested_offset_union
        self.assertEqual(
            self.api.NestedOffsetUnion.CHOICE_NESTED_OFFSET_ARRAY_STRUCTURE, nestedOffsetUnion.choice_tag
        )

        nestedOffsetArrayStructure = nestedOffsetUnion.nested_offset_array_structure
        self.assertEqual(self.NUM_ELEMENTS, nestedOffsetArrayStructure.num_elements)

        nestedOffsetStructureList = nestedOffsetArrayStructure.nested_offset_structure_list
        self.assertEqual(self.NUM_ELEMENTS, len(nestedOffsetStructureList))
        for i in range(self.NUM_ELEMENTS):
            nestedOffsetStructure = nestedOffsetStructureList[i]
            self.assertEqual(self.FIRST_DATA_OFFSET + i * 8, nestedOffsetStructure.data_offset)
            self.assertEqual(i, nestedOffsetStructure.data)

        self.assertEqual(self.TERMINATOR_VALUE, nestedOffset.terminator)

    def _createNestedOffset(self, createWrongOffsets):
        nestedOffsetStructureList = []
        for i in range(self.NUM_ELEMENTS):
            dataOffset = self.WRONG_DATA_OFFSET if createWrongOffsets else self.FIRST_DATA_OFFSET + i * 8
            nestedOffsetStructureList.append(self.api.NestedOffsetStructure(dataOffset, i))

        nestedOffsetArrayStructure = self.api.NestedOffsetArrayStructure(
            self.NUM_ELEMENTS, nestedOffsetStructureList
        )

        nestedOffsetUnion = self.api.NestedOffsetUnion(
            nested_offset_array_structure_=nestedOffsetArrayStructure
        )

        nestedOffsetChoice = self.api.NestedOffsetChoice(
            self.BOOL_VALUE, nested_offset_union_=nestedOffsetUnion
        )

        terminatorOffset = self.WRONG_TERMINATOR_OFFSET if createWrongOffsets else self.TERMINATOR_OFFSET
        nestedOffset = self.api.NestedOffset(
            terminatorOffset, self.BOOL_VALUE, nestedOffsetChoice, self.TERMINATOR_VALUE
        )

        return nestedOffset

    BOOL_VALUE = True
    NUM_ELEMENTS = 2

    WRONG_TERMINATOR_OFFSET = 0
    TERMINATOR_OFFSET = 7 + 2 * 8

    WRONG_DATA_OFFSET = 0
    FIRST_DATA_OFFSET = 7 + 4

    TERMINATOR_VALUE = 0x45

    NEST_OFFSET_BIT_SIZE = (7 + 2 * 8) * 8 + 7
